Libplanet/State/AccountStateDeltaImpl.cs

```diff
@@ -339,53 +340,118 @@ public IAccountStateDelta SetValidator(Validator validator)
-                    accountBalanceGetter,
-                    totalSupplyGetter,
-                    validatorSetGetter,
+                    delta.GetStates,
+                    delta.GetBalance,
+                    delta.GetTotalSupply,
+                    delta.GetValidatorSet,
+                    default(Address))
+                : new AccountStateDeltaImplV0(
+                    delta.GetStates,
+                    delta.GetBalance,
+                    delta.GetTotalSupply,
+                    delta.GetValidatorSet,
+                    default(Address));
+
+        /// <summary>
+        /// Creates a null delta with given <paramref name="signer"/>.
+        /// </summary>
+        /// <param name="delta">The previous <see cref="IAccountStateDelta"/> to use.</param>
+        /// <param name="signer">The <see cref="Address"/> to set.</param>
+        /// <returns>A null delta with given <paramref name="signer"/> that is of the same type
+        /// as <paramref name="delta"/>.</returns>
+        /// <exception cref="ArgumentException">Thrown if given <paramref name="delta"/>
+        /// is neither <see cref="AccountStateDeltaImplV0"/>
+        /// nor <see cref="AccountStateDeltaImpl"/>.
+        /// </exception>
+        /// <remarks>
+        /// This inherits <paramref name="delta"/>'s
+        /// <see cref="IAccountStateDelta.TotalUpdatedFungibleAssets"/>.
+        /// </remarks>
+        internal static IAccountStateDelta ChooseSigner(
+            IAccountStateDelta delta,
+            Address signer)
+        {
+            // The order is important since AccountStateDeltaImplV0 inherits from
+            // AccountStateDeltaImpl
+            if (delta is AccountStateDeltaImplV0 implV0)
+            {
+                return new AccountStateDeltaImplV0(
+                    delta.GetStates,
+                    delta.GetBalance,
+                    delta.GetTotalSupply,
+                    delta.GetValidatorSet,
                     signer)
                     {
-                        TotalUpdatedFungibles = previousTotalUpdatedFungibles,
-                    }
-                : new AccountStateDeltaImplV0(
-                    accountStateGetter,
-                    accountBalanceGetter,
-                    totalSupplyGetter,
-                    validatorSetGetter,
+                        TotalUpdatedFungibles = implV0.TotalUpdatedFungibles,
+                    };
+            }
+            else if (delta is AccountStateDeltaImpl impl)
+            {
+                return new AccountStateDeltaImpl(
+                    delta.GetStates,
+                    delta.GetBalance,
+                    delta.GetTotalSupply,
+                    delta.GetValidatorSet,
                     signer)
                     {
-                        TotalUpdatedFungibles = previousTotalUpdatedFungibles,
+                        TotalUpdatedFungibles = impl.TotalUpdatedFungibles,
                     };
```

Comment:
IMO using a `switch statement` is much better than an `if statement` in this usage.

Reply:
Fixed. 🙂